src/core/reference/include/openvino/reference/paged_attention.hpp

```diff
@@ -0,0 +1,330 @@
+                        } else {
+                            score_val = paged_attention_utils::dot_product(q_vec, key_vec, head_size);
+                        }
+                    }
+                } else {
+                    // Retrieve key from new input.
+                    int32_t new_token_idx = subsequence_begins ? (subsequence_begins[seq_idx] + (k - seq_past_tokens))
+                                                               : (k - seq_past_tokens);
+                    int32_t kv_head = h % num_kv_heads;
+                    const T* key_vec = key + new_token_idx * kv_shape[1] + kv_head * head_size;
+                    score_val = paged_attention_utils::dot_product(q_vec, key_vec, head_size);
+                }
+                // Apply scale and add alibi bias.
+                score_val *= scale;
+                T alibi = alibi_slopes ? alibi_slopes[h % num_kv_heads] : T(0);
+                score_val += alibi * static_cast<T>(k);
+                scores[k] = score_val;
+            }
+
+            paged_attention_utils::softmax(scores);
+
+            // Compute weighted sum over value vectors.
+            std::vector<T> out_vec(head_size, T(0));
+            for (int32_t k = 0; k < total_keys; k++) {
+                T weight = scores[k];
+                if (k < seq_past_tokens) {
+                    int32_t block_id = -1, token_offset = 0;
+                    bool found = paged_attention_utils::find_cached_token(seq_idx,
+                                                                          k,
+                                                                          block_indices,
+                                                                          block_indices_begins,
+                                                                          num_blocks,
+                                                                          block_size,
+                                                                          block_id,
+                                                                          token_offset);
+                    if (!found)
+                        continue;
+                    // Skip accumulation for tokens in the sliding window region.
+                    int32_t first_block_for_seq =
+                        (block_indices_begins ? block_indices[block_indices_begins[seq_idx]] : -1);
+                    if (first_block_for_seq >= 0 && block_id == first_block_for_seq && token_offset < sliding_window)
+                        continue;
+
+                    int32_t kv_head = h % num_kv_heads;
+                    const T* raw_val_vec =
+                        value_cache + (((block_id * num_kv_heads + kv_head) * block_size + token_offset) * head_size);
+
+                    bool do_rotate = false;
+                    int32_t rotated_index = -1;
+                    if (rotated_block_indices && rotation_deltas && rotation_trig_lut) {
```

Comment:
Can we agree that all the necessary checks will be implemented inside the helper functions, and they will return `false` in case of invalid arguments?  I hope it will simplify the outer code

Reply:
Done